### PR TITLE
Convert user_access_token_section to TypeScript

### PR DESCRIPTION
**Pull request by hiendinhngoc**
## Summary
Migrate 'components/user_settings/security/user_access_token_section' module and associated tests to TypeScript

#### Ticket Link

  Fixes https://github.com/mattermost/mattermost-server/issues/13693
  JIRA: https://mattermost.atlassian.net/browse/MM-20548

### CODE REVIEW

**Comment by hiendinhngoc** [comment]
I got an error at line 50 in `index.ts` file:
`userAccessTokens: state.entities.users.myUserAccessTokens`

error: `Property 'myUserAccessTokens' does not exist on type 'UsersState'.`
I already checked [UsersState](https://github.com/mattermost/mattermost-redux/blob/62d85b6194751d88b2474d86a1019b683f8b1340/src/types/users.ts#L59), it really doesn't have `myUserAccessTokens` property.

**Comment by devinbinnie** [comment]
> I got an error at line 50 in `index.ts` file:
> `userAccessTokens: state.entities.users.myUserAccessTokens`
> 
> error: `Property 'myUserAccessTokens' does not exist on type 'UsersState'.`
> I already checked [UsersState](https://github.com/mattermost/mattermost-redux/blob/62d85b6194751d88b2474d86a1019b683f8b1340/src/types/users.ts#L59), it really doesn't have `myUserAccessTokens` property.

Thanks for the catch! I'll make a change to redux and get that fixed.

**Comment by devinbinnie** [comment]
@hiendinhngoc The fix is in, you'll just need to update the `mattermost-redux` package to this commit and everything should work nicely: `1838b6c20c51d9f2a8c9899c91417672dced87f0`

**Comment by hiendinhngoc** [comment]
> @hiendinhngoc The fix is in, you'll just need to update the `mattermost-redux` package to this commit and everything should work nicely: `1838b6c20c51d9f2a8c9899c91417672dced87f0`

@devinbinnie thank you, I updated in my side by entering the version to `package.json` file, then `npm install`. After that, the error came out and type-check was passed from my local. However, CI shows the error with type-check. Is there any step that I missed or sth wrong here?

**Comment by devinbinnie** [comment]
> > @hiendinhngoc The fix is in, you'll just need to update the `mattermost-redux` package to this commit and everything should work nicely: `1838b6c20c51d9f2a8c9899c91417672dced87f0`
> 
> @devinbinnie thank you, I updated in my side by entering the version to `package.json` file, then `npm install`. After that, the error came out and type-check was passed from my local. However, CI shows the error with type-check. Is there any step that I missed or sth wrong here?

@hiendinhngoc CircleCI uses the copy of `package-lock.json` in the repo in order to fetch its packages, so you'll need to change that as well. Sorry for the inconvience.

**Comment by hiendinhngoc** [comment]
@devinbinnie It seems an update for mattermost-redux at [previous commit](https://github.com/mattermost/mattermost-redux/commit/285e4105b21fcbe8ff091d302b92a9f9a8e0b262) which caused type check error when I changed to `1838b6c20c51d9f2a8c9899c91417672dced87f0`

**Comment by devinbinnie** [comment]
> @devinbinnie It seems an update for mattermost-redux at [previous commit](https://github.com/mattermost/mattermost-redux/commit/285e4105b21fcbe8ff091d302b92a9f9a8e0b262) which caused type check error when I changed to `1838b6c20c51d9f2a8c9899c91417672dced87f0`

@hiendinhngoc I guess you'll have to update the files to the latest commit then, see here: https://github.com/mattermost/mattermost-redux/commit/6fdb4667fdb689ac209eb27d2726922d8b422104

**Comment by jasonblais** [comment]
Hi @hiendinhngoc, did you have questions on how to update the files to the latest commit: mattermost/mattermost-redux@6fdb466

If any, let us know! Happy to help :)

File: package-lock.json [hunk 1]
```diff
@@ -6936,8 +6936,7 @@
           "version": "3.0.0",
           "resolved": "https://registry.npmjs.org/get-stream/-/get-stream-3.0.0.tgz",
           "integrity": "sha1-jpQ9E1jcN1VQVOy+LtsFqhdO3hQ=",
-          "dev": true,
-          "optional": true
+          "dev": true
```
Comment (devinbinnie) [inline package-lock.json:6939]: Can we revert these `package-lock.json` changes? They aren't relevant to this PR.

File: components/user_settings/security/user_access_token_section/user_access_token_section.tsx [hunk 2]
```diff
@@ -296,19 +328,19 @@ export default class UserAccessTokenSection extends React.PureComponent {
...
 
-            let activeLink;
+            let activeLink: JSX.Element;
             let activeStatus;
 
             if (token.is_active) {
                 activeLink = (
                     <a
-                        name={token.id + '_deactivate'}
+                        data-name={token.id + '_deactivate'}
```
Comment (devinbinnie) [inline components/user_settings/security/user_access_token_section/user_access_token_section.tsx:343]: Why was this attribute name changed?

**Review by devinbinnie — COMMENTED** [review]
Looks good overall @hiendinhngoc, thanks for the contribution!
Just had a couple comments.

File: components/user_settings/security/user_access_token_section/user_access_token_section.tsx (hunk 2 above)
Comment (hiendinhngoc) [reply components/user_settings/security/user_access_token_section/user_access_token_section.tsx:343]: > Why was this attribute name changed?

I got error: 
```js
Type '{ children: Element; name: string; href: string; onClick: (e: MouseEvent<HTMLAnchorElement, MouseEvent>) => void; }' is not assignable to type 'DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>'.
  Property 'name' does not exist on type 'DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>'.

(JSX attribute) name: string
```
so I fixed this by adding `data-*`

Comment (devinbinnie) [reply components/user_settings/security/user_access_token_section/user_access_token_section.tsx:343]: Hmm, looks like it should probably be an `id`. Can we change it to that?

Comment (hiendinhngoc) [reply components/user_settings/security/user_access_token_section/user_access_token_section.tsx]: I changed, and it works, thank you. There is an issue, I ran type check in my local and it passed, however, CI shows the error at the unrelated files. https://app.circleci.com/pipelines/github/mattermost/mattermost-webapp/28734/workflows/8bc8969c-bec9-44ed-8b12-b918b38427bc/jobs/189515

Comment (devinbinnie) [reply components/user_settings/security/user_access_token_section/user_access_token_section.tsx]: @hiendinhngoc Just had to update your branch, looks like checks are passing now :)

**Review by devinbinnie — APPROVED** [review]
LGTM! Thanks @hiendinhngoc!

File: components/user_settings/security/user_access_token_section/index.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    enableUserAccessToken,
+    disableUserAccessToken,
+} from 'mattermost-redux/actions/users';
+import {GlobalState} from 'mattermost-redux/types/store';
+import {ActionFunc} from 'mattermost-redux/types/actions';
+
+import UserAccessTokenSection from './user_access_token_section';
+
+type Actions = {
+    getUserAccessTokensForUser: (userId: string, num1: number, num2: number) => void;
```
Comment (hmhealey) [inline components/user_settings/security/user_access_token_section/index.ts:20]:
```suggestion
    getUserAccessTokensForUser: (userId: string, page: number, perPage: number) => void;
```

File: components/user_settings/security/user_access_token_section/user_access_token_section.tsx [hunk 4]
```diff
@@ -23,29 +23,61 @@ const TOKEN_CREATING = 'creating';
...
-        }).isRequired,
-    }
+type Props = {
+    user: UserProfile;
+    active?: boolean;
+    updateSection: (section: string) => void;
+    userAccessTokens: {[tokenId: string]: {description: string; id: string; is_active: boolean}};
+    setRequireConfirm: (isRequiredConfirm: boolean, confirmCopyToken: (confirmAction: () => void) => void) => void;
+    actions: {
+        getUserAccessTokensForUser: (userId: string, num1: number, num2: number) => void;
```
Comment (hmhealey) [inline components/user_settings/security/user_access_token_section/user_access_token_section.tsx:33]:
```suggestion
        getUserAccessTokensForUser: (userId: string, page: number, perPage: number) => void;
```

File: components/user_settings/security/user_access_token_section/user_access_token_section.tsx [hunk 5]
```diff
@@ -23,29 +23,61 @@ const TOKEN_CREATING = 'creating';
...
+    active?: boolean;
+    showConfirmModal: boolean;
+    newToken?: {token: string; description: string; id: string; is_active: boolean} | null;
+    tokenCreationState?: string;
+    tokenError?: string;
+    serverError?: string|null;
+    saving?: boolean;
+    confirmTitle?: JSX.Element|null;
+    confirmMessage?: ((state: State) => JSX.Element)|null;
+    confirmButton?: JSX.Element|null;
```
Comment (hmhealey) [inline components/user_settings/security/user_access_token_section/user_access_token_section.tsx:72]:
```suggestion
    confirmTitle: React.ReactNode;
    confirmMessage?: ((state: State) => JSX.Element)|null;
    confirmButton: React.ReactNode;
```

**Review by hmhealey — CHANGES_REQUESTED** [review]
Thanks for the work on this, @hiendinhngoc. I have a couple suggestions for things to change, but they're just tightening up the type definitions slightly.

**Review by hmhealey — APPROVED** [review]
Thanks! This is good for QA review then

**Review by jgilliam17 — APPROVED** [review]
Thank you @hiendinhngoc 
Tested, looks good to merge.
- Verified Personal Access Tokens section in Account Settings>Security. Able to create, deactivate, delete - as expected.
@devinbinnie Can you please help merge? Thanks. 🙂

**Pull request merged**